### PR TITLE
[aot_autograd] Deprecate backward_pass_autocast="same_as_forward" default

### PR DESCRIPTION
**Pull request by aorenste**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #177287
* __->__ #177286

The current default of `backward_pass_autocast="same_as_forward"` causes
the backward pass of `torch.compile`'d regions to inherit the forward's
autocast context. This doesn't match eager PyTorch behavior, where autocast
conventionally wraps only the forward pass, and can cause silent numerical
issues like gradient overflow (#153044).

This change introduces a deprecation warning (FutureWarning) for users who
have autocast active during compilation without explicitly setting
`backward_pass_autocast`. The default config value is changed to `"default"`,
which currently behaves like `"same_as_forward"` but will change to `"off"`
in a future release.

To silence the warning, users should set:
  torch._functorch.config.backward_pass_autocast = "off"   # recommended
  torch._functorch.config.backward_pass_autocast = "same_as_forward"

## Release note:
The default behavior of `backward_pass_autocast` in
`torch._functorch.config` is deprecated and will change in a future release.
Currently, the backward pass of `torch.compile`'d regions inherits the
forward pass's autocast context (`"same_as_forward"`). In a future release,
the default will change to `"off"`, matching eager PyTorch behavior where
autocast wraps only the forward pass. Users who rely on the current behavior
should explicitly set `torch._functorch.config.backward_pass_autocast =
"same_as_forward"`. Users who want the new behavior now can set it to `"off"`.

[bc-breaking][deprecation]

Authored with Claude.

cc @ezyang @gchanan

### CODE REVIEW

File: docs/source/user_guide/torch_compiler/torch.compiler_backward.md [hunk 1]
```diff
@@ -30,8 +30,17 @@ Use ``torch._functorch.config.backward_pass_autocast``
 to control that assumption; an incorrect assumption may lead to silent
 incorrectness.
 
+```{note}
+The default value of `backward_pass_autocast` will change from
+`"same_as_forward"` to `"off"` in a future release. `"off"` matches the
+recommended PyTorch usage where autocast wraps only the forward pass.
+If you rely on `"same_as_forward"` behavior, set it explicitly to silence
+the deprecation warning.
+See [#153044](https://github.com/pytorch/pytorch/issues/153044) for details.
```
Comment (zou3519) [inline docs/source/user_guide/torch_compiler/torch.compiler_backward.md:39]: link https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/torch.compiler_backward.html#autocast-behavior instead of an issue

File: docs/source/user_guide/torch_compiler/torch.compiler_backward.md [hunk 2]
```diff
@@ -30,8 +30,17 @@ Use ``torch._functorch.config.backward_pass_autocast``
 to control that assumption; an incorrect assumption may lead to silent
 incorrectness.
 
+```{note}
+The default value of `backward_pass_autocast` will change from
+`"same_as_forward"` to `"off"` in a future release. `"off"` matches the
+recommended PyTorch usage where autocast wraps only the forward pass.
+If you rely on `"same_as_forward"` behavior, set it explicitly to silence
```
Comment (zou3519) [inline docs/source/user_guide/torch_compiler/torch.compiler_backward.md:37]: tell the user what to set. e.g. `add torch._functorch.config.backward_pass_autocast = blah to the top of your file`

File: torch/_functorch/_aot_autograd/graph_capture_wrappers.py [hunk 3]
```diff
@@ -416,6 +416,22 @@ def inner_fn(
                 ExitStack() as stack,
             ):
                 backward_pass_autocast = torch._functorch.config.backward_pass_autocast
+                if backward_pass_autocast == "default":
```
Comment (zou3519) [inline torch/_functorch/_aot_autograd/graph_capture_wrappers.py:419]: This is going to be very spammy. Basically everyone who calls torch.compile with requires_grad=True tensors will see this warning.

Brian and I discussed this before, and we thought it would be better to "only warn if it matters".

**Review by zou3519 — COMMENTED** [review]
My only concern is really just the spamminess. I haven't worked out if we can spam people less.

**Review by zou3519 — COMMENTED** [review]
@aorenste can you figure out a strategy to spam people less with the warning?

Otherwise I'd prefer pre-existing silent incorrectness to spamming people
